Increase the Entropy of ModuleManager Map Keys

CodaFi opened the pull request:
The ModuleManager's use of FileEntry nodes as the keys for its map of
loaded modules is less than ideal. Uniqueness for FileEntry nodes is
maintained by FileManager, which in turn uses inode numbers on hosts
that support that. When coupled with the module cache's proclivity for
turning over and deleting stale PCMs, this means entries for different
module files can wind up reusing the same underlying inode. When this
happens, subsequent accesses to the Modules map will disagree on the
ModuleFile associated with a given file.

It's fine to use the file management utilities to guarantee the presence
of module data, but we need a better source of key material that is
invariant with respect to these OS-level details. Using file paths alone
is a similarly frought solution because the ASTWriter performs a custom
canonicalization step (that is not equivalent to path canonicalization)
that renders keys from loaded AST cores useless for looking up cached
entries.

To mitigate the effects of inode reuse, increase the entropy of the key
material by incorporating the modtime and size. This ultimately
decreases the likelihood that a PCM that is swapped on disk will confuse
the cache, but it does not eliminate the possibility of collisions.

rdar://48443680

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci clean test macOS

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test

hyp commented:
@CodaFi Swift/master branch only accepts LLDB changes. Please merge your PR into `apple/stable/20200108` instead , and also cherry-pick it to `apple/stable/20200714`.

CodaFi commented:
@hyp I'll bear that in mind. Using this to kick off one last set of tests.

@swift-ci test

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test

CodaFi committed:
Increase the Entropy of ModuleManager Map Keys

The ModuleManager's use of FileEntry nodes as the keys for its map of
loaded modules is less than ideal. Uniqueness for FileEntry nodes is
maintained by FileManager, which in turn uses inode numbers on hosts
that support that. When coupled with the module cache's proclivity for
turning over and deleting stale PCMs, this means entries for different
module files can wind up reusing the same underlying inode. When this
happens, subsequent accesses to the Modules map will disagree on the
ModuleFile associated with a given file.

It's fine to use the file management utilities to guarantee the presence
of module data, but we need a better source of key material that is
invariant with respect to these OS-level details. Using file paths alone
is a similarly frought solution because the ASTWriter performs a custom
canonicalization step (that is not equivalent to path canonicalization)
that renders keys from loaded AST cores useless for looking up cached
entries.

To mitigate the effects of inode reuse, increase the entropy of the key
material by incorporating the modtime and size. This ultimately
decreases the likelihood that a PCM that is swapped on disk will confuse
the cache, but it does not eliminate the possibility of collisions.

rdar://48443680

CodaFi commented:
@swift-ci test

CodaFi commented:
⛵
